File: dev/src/samples/NetSamplesVB/Samples/CogoPoints/CogoPointLabelCommands.vb
Imports System.Collections.Generic

Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.EditorInput
Imports Autodesk.AutoCAD.Runtime
Imports Autodesk.Civil
Imports Autodesk.Civil.DatabaseServices
Imports Autodesk.Civil.DatabaseServices.Styles

<Assembly: CommandClass(
  GetType(Autodesk.CivilizedDevelopment.CogoPointLabelCommands))> 

Namespace Autodesk.CivilizedDevelopment
  Class CogoPointLabelCommands
    Inherits SimpleDrawingCommand
    <CommandMethod("CDS_CreateDemoPointLabelStyle")> _
    Public Sub CDS_CreateDemoPointLabelStyle()

      createPointLabelStyle("Demo")
    End Sub

    Private Sub createPointLabelStyle(name As String)
      Dim styleId As ObjectId = _pointLabelStyles.Add(name)
      removeAllComponents(styleId)
      customizeStyle(styleId)
    End Sub

    Private ReadOnly Property _pointLabelStyles() As LabelStyleCollection
      Get
        Return _civildoc.Styles.LabelStyles.PointLabelStyles.LabelStyles
      End Get
    End Property

    Private Sub removeAllComponents(styleId As ObjectId)
      Dim componentNames As IEnumerable(Of String) = getTextComponentNames(
        styleId)
      removeComponents(styleId, componentNames)
    End Sub

    Private Function getTextComponentNames(styleId As ObjectId) _
        As IEnumerable(Of String)
      Dim names As New List(Of String)()
      Using tr As Transaction = startTransaction()
        Dim style As LabelStyle = TryCast(styleId.GetObject(OpenMode.ForRead), 
          LabelStyle)
        For Each id As ObjectId In style.GetComponents(
            LabelStyleComponentType.Text)
          Dim component As LabelStyleComponent = TryCast(
            id.GetObject(OpenMode.ForRead), LabelStyleComponent)
          names.Add(component.Name)
        Next
      End Using
      Return names
    End Function

    Private Sub removeComponents(styleId As ObjectId, componentNames _
        As IEnumerable(Of String))
      Using tr As Transaction = startTransaction()
        Dim style As LabelStyle = TryCast(styleId.GetObject(OpenMode.ForWrite), 
          LabelStyle)
        For Each name As String In componentNames
          style.RemoveComponent(name)
        Next

        tr.Commit()
      End Using
    End Sub

    Private Sub customizeStyle(styleId As ObjectId)
      Using tr As Transaction = startTransaction()
        addStyleComponents(styleId)
        tr.Commit()
      End Using
    End Sub

    Private Sub addStyleComponents(styleId As ObjectId)
      Dim style As LabelStyle = TryCast(styleId.GetObject(OpenMode.ForWrite), 
        LabelStyle)
      addLeaderComponent(style)
      addPointNumberComponent(style)
      addLocationComponent(style)
    End Sub

    Private Sub addLeaderComponent(style As LabelStyle)
      Dim id As ObjectId = style.AddComponent("Leader",
        LabelStyleComponentType.Line)
      Dim component As LabelStyleLineComponent = TryCast(
        id.GetObject(OpenMode.ForWrite), LabelStyleLineComponent)
      component.General.StartAnchorPoint.Value = AnchorPointType.MiddleCenter
    End Sub

    Private Sub addPointNumberComponent(style As LabelStyle)
      Dim id As ObjectId = style.AddComponent("PN",
        LabelStyleComponentType.Text)
      Dim component As LabelStyleTextComponent = TryCast(
        id.GetObject(OpenMode.ForWrite), LabelStyleTextComponent)
      component.Text.Attachment.Value = LabelTextAttachmentType.MiddleLeft
      component.Text.Contents.Value = _pointNumber
      component.General.AnchorComponent.Value = "Leader"
      component.General.AnchorLocation.Value = AnchorPointType.[End]
    End Sub

    Private Sub addLocationComponent(style As LabelStyle)
      Dim id As ObjectId = style.AddComponent("Location",
        LabelStyleComponentType.Text)
      Dim component As LabelStyleTextComponent = TryCast(
        id.GetObject(OpenMode.ForWrite), LabelStyleTextComponent)
      component.Text.Attachment.Value = LabelTextAttachmentType.TopLeft
      Dim value As String = [String].Format("({0}, {1}, {2})",
        _northing, _easting, _elevation)
      component.Text.Contents.Value = value
      component.General.AnchorComponent.Value = "PN"
      component.General.AnchorLocation.Value = AnchorPointType.BottomLeft
    End Sub

    <CommandMethod("CDS_ShowCogoPointLabelProperties")> _
    Public Sub CDS_ShowCogoPointLabelProperties()
      Dim pointId As ObjectId = selectCogoPoint()
      If pointId = ObjectId.Null Then
        Return
      End If
      showLabelProperties(pointId)
    End Sub

    <CommandMethod("CDS_OverrideLabelsForGroup")> _
    Public Sub CDS_OverrideLabelsForGroup()
      Using tr As Transaction = startTransaction()
        Dim pointGroupId As ObjectId = promptForPointGroup()
        If pointGroupId = ObjectId.Null Then
          Return
        End If
        overrideLabelsForPointsIn(pointGroupId)
        tr.Commit()
      End Using

    End Sub

    Private Function selectCogoPoint() As ObjectId
      Dim options As New PromptEntityOptions("Select COGO point: ")
      options.SetRejectMessage(vbLf & "Invalid COGO point selected.")
      options.AddAllowedClass(GetType(CogoPoint), True)
      Dim result As PromptEntityResult = _editor.GetEntity(options)
      If result.Status = PromptStatus.OK Then
        Return result.ObjectId
      End If
      Return ObjectId.Null
    End Function

    Private Sub showLabelProperties(pointId As ObjectId)
      Using tr As Transaction = startTransaction()
        Dim point As CogoPoint = TryCast(pointId.GetObject(OpenMode.ForRead), 
          CogoPoint)
        showLabelPropertiesFor(point)
      End Using
    End Sub

    Private Sub showLabelPropertiesFor(point As CogoPoint)
      write(vbLf & "Point Label Properties:")
      write(vbLf & "- Style: " & getLabelStyleName(point.LabelStyleId))
      write(vbLf & "- Style override: " & getLabelStyleName(
        point.LabelStyleIdOverride))
      write(vbLf & "- Visible: " & point.IsLabelVisible.ToString())
      write(vbLf & "- Location: " & point.LabelLocation.ToString())
      write(vbLf & "- Rotation: " & point.LabelRotation.ToString())
      write(vbLf & "- Dragged: " & point.IsLabelDragged.ToString())
      write(vbLf & "- Pinned: " & point.IsLabelPinned.ToString())
    End Sub

    Private Function getLabelStyleName(id As ObjectId) As String
      Dim style As LabelStyle = TryCast(id.GetObject(OpenMode.ForRead), 
        LabelStyle)
      Return style.Name
    End Function

    Private Function promptForPointGroup() As ObjectId
      Dim result As PromptResult = _editor.GetString(
        vbLf & "Enter point group name: ")
      If result.Status = PromptStatus.OK Then
        Return findGroup(result.StringResult)
      End If
      Return ObjectId.Null
    End Function

    Private Sub overrideLabelsForPointsIn(pointGroupId As ObjectId)
      Dim group As PointGroup = TryCast(pointGroupId.GetObject(OpenMode.ForWrite), 
        PointGroup)
      group.IsPointLabelStyleOverridden = True
    End Sub

    Private Function findGroup(name As String) As ObjectId
      For Each id As ObjectId In _civildoc.PointGroups
        Dim group As PointGroup = TryCast(id.GetObject(OpenMode.ForRead), 
          PointGroup)
        If group.Name = name Then
          Return id
        End If
      Next
      Return ObjectId.Null
    End Function

    Private ReadOnly _pointNumber As String = "<[Point Number(Sn)]>"
    Private ReadOnly _northing As String =
      "<[Northing(Uft|P4|RN|AP|GC|UN|Sn|OF)]>"
    Private ReadOnly _easting As String =
      "<[Easting(Uft|P4|RN|AP|GC|UN|Sn|OF)]>"
    Private ReadOnly _elevation As String =
      "<[Point Elevation(Uft|P2|RN|AP|Sn|OF)]>"
  End Class
End Namespace

File: dev/src/samples/NetSamples/Samples/CogoPoints/CogoPointLabelCommands.cs
using System;
using System.Collections.Generic;

using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.EditorInput;
using Autodesk.AutoCAD.Runtime;
using Autodesk.Civil;
using Autodesk.Civil.DatabaseServices;
using Autodesk.Civil.DatabaseServices.Styles;

[assembly: CommandClass(
    typeof(Autodesk.CivilizedDevelopment.CogoPointLabelCommands))]

namespace Autodesk.CivilizedDevelopment
{
    class CogoPointLabelCommands : SimpleDrawingCommand
    {
        [CommandMethod("CDS_CreateDemoPointLabelStyle")]
        public void CDS_CreateDemoPointLabelStyle()
        {
            
            createPointLabelStyle("Demo"); 
        }

        private void createPointLabelStyle(string name)
        {
            ObjectId styleId = _pointLabelStyles.Add(name);
            removeAllComponents(styleId);
            customizeStyle(styleId);
        }

        private LabelStyleCollection _pointLabelStyles
        {
            get
            {
                return _civildoc.Styles.LabelStyles.PointLabelStyles.LabelStyles;
            }
        }

        private void removeAllComponents(ObjectId styleId)
        {
            IEnumerable<string> componentNames = getTextComponentNames(styleId);
            removeComponents(styleId, componentNames);
        }

        private IEnumerable<string> getTextComponentNames(ObjectId styleId)
        {
            List<string> names = new List<string>();
            using (Transaction tr = startTransaction())
            {
                LabelStyle style = styleId.GetObject(OpenMode.ForRead)
                    as LabelStyle;
                foreach (ObjectId id in style.GetComponents(
                    LabelStyleComponentType.Text))
                {
                    LabelStyleComponent component =
                        id.GetObject(OpenMode.ForRead) as LabelStyleComponent;
                    names.Add(component.Name);
                }
            }
            return names;
        }

        private void removeComponents(ObjectId styleId, 
            IEnumerable<string> componentNames)
        {
            using (Transaction tr = startTransaction())
            {
                LabelStyle style = styleId.GetObject(OpenMode.ForWrite)
                    as LabelStyle;
                foreach (string name in componentNames)
                {
                    style.RemoveComponent(name);
                }

                tr.Commit();
            }
        }

        private void customizeStyle(ObjectId styleId)
        {
            using (Transaction tr = startTransaction())
            {
                addStyleComponents(styleId);
                tr.Commit();
            }
        }

        private void addStyleComponents(ObjectId styleId)
        {
            LabelStyle style = styleId.GetObject(OpenMode.ForWrite)
                    as LabelStyle;
            addLeaderComponent(style);
            addPointNumberComponent(style);
            addLocationComponent(style);
        }

        private void addLeaderComponent(LabelStyle style)
        {
            ObjectId id = style.AddComponent("Leader", 
                LabelStyleComponentType.Line);
            LabelStyleLineComponent component = id.GetObject(OpenMode.ForWrite)
                as LabelStyleLineComponent;
            component.General.StartAnchorPoint.Value = AnchorPointType.MiddleCenter;
        }

        private void addPointNumberComponent(LabelStyle style)
        {
            ObjectId id = style.AddComponent("PN", 
                LabelStyleComponentType.Text);
            LabelStyleTextComponent component = id.GetObject(OpenMode.ForWrite)
                as LabelStyleTextComponent;
            component.Text.Attachment.Value = LabelTextAttachmentType.MiddleLeft;
            component.Text.Contents.Value = _pointNumber;
            component.General.AnchorComponent.Value = "Leader";
            component.General.AnchorLocation.Value = AnchorPointType.End;
        }

        private void addLocationComponent(LabelStyle style)
        {
            ObjectId id = style.AddComponent("Location", 
                LabelStyleComponentType.Text);
            LabelStyleTextComponent component = id.GetObject(OpenMode.ForWrite)
                as LabelStyleTextComponent;
            component.Text.Attachment.Value = LabelTextAttachmentType.TopLeft;
            string value = String.Format("({0}, {1}, {2})", 
                _northing, _easting, _elevation);
            component.Text.Contents.Value = value;
            component.General.AnchorComponent.Value = "PN";
            component.General.AnchorLocation.Value = AnchorPointType.BottomLeft;
        }

        [CommandMethod("CDS_ShowCogoPointLabelProperties")]
        public void CDS_ShowCogoPointLabelProperties()
        {
            ObjectId pointId = selectCogoPoint();
            if (pointId == ObjectId.Null)
            {
                return;
            }
            showLabelProperties(pointId);
        }

        [CommandMethod("CDS_OverrideLabelsForGroup")]
        public void CDS_OverrideLabelsForGroup()
        {
            using (Transaction tr = startTransaction())
            {
                ObjectId pointGroupId = promptForPointGroup();
                if (pointGroupId == ObjectId.Null)
                {
                    return;
                }
                overrideLabelsForPointsIn(pointGroupId);
                tr.Commit();
            }
            
        }

        private ObjectId selectCogoPoint()
        {
            PromptEntityOptions options = new PromptEntityOptions(
                "Select COGO point: ");
            options.SetRejectMessage("\nInvalid COGO point selected.");
            options.AddAllowedClass(typeof(CogoPoint), true);
            PromptEntityResult result = _editor.GetEntity(options);
            if (result.Status == PromptStatus.OK)
            {
                return result.ObjectId;
            }
            return ObjectId.Null;
        }

        private void showLabelProperties(ObjectId pointId)
        {
            using (Transaction tr = startTransaction())
            {
                CogoPoint point = pointId.GetObject(OpenMode.ForRead)
                    as CogoPoint;
                showLabelPropertiesFor(point);
            }
        }

        private void showLabelPropertiesFor(CogoPoint point)
        {
            write("\nPoint Label Properties:");
            write("\n- Style: " + getLabelStyleName(point.LabelStyleId));
            write("\n- Style override: " 
                + getLabelStyleName(point.LabelStyleIdOverride));
            write("\n- Visible: " + point.IsLabelVisible.ToString());
            write("\n- Location: " + point.LabelLocation.ToString());
            write("\n- Rotation: " + point.LabelRotation.ToString());
            write("\n- Dragged: " + point.IsLabelDragged.ToString());
            write("\n- Pinned: " + point.IsLabelPinned.ToString());
        }

        private string getLabelStyleName(ObjectId id)
        {
            LabelStyle style = id.GetObject(OpenMode.ForRead) as LabelStyle;
            return style.Name;
        }

        private ObjectId promptForPointGroup()
        {
            PromptResult result = _editor.GetString(
                "\nEnter point group name: ");
            if (result.Status == PromptStatus.OK)
            {
                return findGroup(result.StringResult);
            }
            return ObjectId.Null;
        }

        private void overrideLabelsForPointsIn(ObjectId pointGroupId)
        {
            PointGroup group = pointGroupId.GetObject(OpenMode.ForWrite)
                as PointGroup;
            group.IsPointLabelStyleOverridden = true;
        }

        private ObjectId findGroup(string name)
        {
            foreach (ObjectId id in _civildoc.PointGroups)
            {
                PointGroup group = id.GetObject(OpenMode.ForRead) 
                    as PointGroup;
                if (group.Name == name)
                {
                    return id;
                }
            }
            return ObjectId.Null;
        }

        private readonly string _pointNumber = "<[Point Number(Sn)]>";
        private readonly string _northing = 
            "<[Northing(Uft|P4|RN|AP|GC|UN|Sn|OF)]>";
        private readonly string _easting =
            "<[Easting(Uft|P4|RN|AP|GC|UN|Sn|OF)]>";
        private readonly string _elevation =
            "<[Point Elevation(Uft|P2|RN|AP|Sn|OF)]>";
    }
}
